File: 07-Repeticion/03-Seguridad/05-Haskell/Seguridad.hs
{-
   Crear un programa para contar cuántas letras mayúsculas,
   cuántas minúsculas y cuántos dígitos contiene una contraseña

   Autor: Julián Esteban Gutiérrez Posada
   Fecha: Septiembre 2025
   Licencia: GNU GPL v3
-}

import Text.Printf (printf)
import Modulo.Util (mostrarMensaje, ingresarTexto)
import Data.Char (isUpper, isLower, isDigit)

main :: IO()
main = do
    contrasenia <- ingresarTexto "Ingrese la contraseña a analizar: "
    
    let cantidadMayusculas = contarMayusculas contrasenia
    let cantidadMinusculas = contarMinusculas contrasenia
    let cantidadDigitos = contarDigitos contrasenia

    let reporteEstres = generarReporteContrasenia contrasenia cantidadMayusculas cantidadMinusculas cantidadDigitos

    mostrarMensaje reporteEstres

contarMayusculas :: String -> Int
contarMayusculas texto = contarCaracteres texto esMayuscula

contarMinusculas :: String -> Int
contarMinusculas texto = contarCaracteres texto esMinuscula

contarDigitos :: String -> Int
contarDigitos texto = contarCaracteres texto esDigito

contarCaracteres :: String -> (Char -> Bool) -> Int
contarCaracteres texto funcion = length (filter funcion texto)

esMayuscula :: Char -> Bool
esMayuscula caracter =  isUpper caracter

esMinuscula :: Char -> Bool
esMinuscula caracter =  isLower caracter

esDigito :: Char -> Bool
esDigito caracter =  isDigit caracter

generarReporteContrasenia :: String -> Int -> Int -> Int -> String
generarReporteContrasenia contrasenia cantidadMayusculas cantidadMinusculas cantidadDigitos =
  printf ("\nEn la constraseña \"%s\" hay:\n" ++
          "%d mayúscula(s), "++
          "%d minúscula(s) y "++
          "%d dígito(s)\n")
          contrasenia cantidadMayusculas cantidadMinusculas cantidadDigitos

File: 07-Repeticion/03-Seguridad/01-Python/Seguridad.py
"""
    Crear un programa para contar cuántas letras mayúsculas, 
    cuántas minúsculas y cuántos dígitos contiene una contraseña

    Autor: Julián Esteban Gutiérrez Posada
    Fecha: Septiembre 2025
    Licencia: GNU GPL v3
"""

from modulo.util import ingresar_texto, mostrar_mensaje

def main():
    contrasenia = ingresar_texto("Ingrese la contraseña a analizar: ") 
    
    cantidad_mayusculas = contar_mayusculas(contrasenia)
    cantidad_minusculas = contar_minusculas(contrasenia)
    cantidad_digitos    = contar_digitos   (contrasenia)
    
    reporte_contrasenia = generar_reporte_contrasenia(contrasenia, 
        cantidad_mayusculas, cantidad_minusculas, cantidad_digitos)

    mostrar_mensaje(reporte_contrasenia)

def contar_mayusculas(texto):
    cantidad_mayusculas = 0

    for caracter in texto:
        if caracter.isupper():
            cantidad_mayusculas += 1

    return cantidad_mayusculas

def contar_minusculas(texto):
    cantidad_minusculas = 0

    for caracter in texto:
        if caracter.islower():
            cantidad_minusculas += 1

    return cantidad_minusculas

def contar_digitos(texto):
    cantidad_digitos = 0

    for caracter in texto:
        if caracter.isdigit():
            cantidad_digitos += 1

    return cantidad_digitos

def generar_reporte_contrasenia(contrasenia, cantidad_mayusculas, cantidad_minusculas, cantidad_digitos):
    return (f"\nEn la constraseña \"{contrasenia}\" hay:\n"
            f"{cantidad_mayusculas} mayúscula(s), "
            f"{cantidad_minusculas} minúscula(s) y "
            f"{cantidad_digitos} dígito(s)\n"
        )

main()